rapier/src/lib.rs

```diff
@@ -70,30 +107,41 @@ enum InternalSystem {
     TransformPropagation,
 }
 
-impl Plugin for RapierPlugin {
+impl<
+        PhysicsSchedule: StageLabel + Clone,
+        PostPhysicsStage: StageLabel + Clone,
+        StepStage: StageLabel + Clone,
+    > Plugin for StagedRapierPlugin<PhysicsSchedule, PostPhysicsStage, StepStage>
+{
     fn build(&self, app: &mut App) {
-        app.add_plugin(heron_core::CorePlugin)
-            .init_resource::<PhysicsPipeline>()
-            .init_resource::<body::HandleMap>()
-            .init_resource::<shape::HandleMap>()
-            .init_resource::<IntegrationParameters>()
-            .add_event::<CollisionEvent>()
-            .insert_resource(BroadPhase::new())
-            .insert_resource(NarrowPhase::new())
-            .insert_resource(RigidBodySet::new())
-            .insert_resource(QueryPipeline::new())
-            .insert_resource(IslandManager::new())
-            .insert_resource(ColliderSet::new())
-            .insert_resource(JointSet::new())
-            .insert_resource(CCDSolver::new())
-            .stage("heron-physics", |schedule: &mut Schedule| {
-                schedule
-                    .add_stage("heron-remove", removal_stage())
-                    .add_stage("heron-update-rapier-world", update_rapier_world_stage())
-                    .add_stage("heron-create-new-bodies", body_update_stage())
-                    .add_stage("heron-create-new-colliders", create_collider_stage())
-            })
-            .add_system_set_to_stage(CoreStage::PostUpdate, step_systems());
+        app.add_plugin(heron_core::StagedCorePlugin {
+            step_stage: self.step_physics_stage.clone(),
+        })
+        .init_resource::<PhysicsPipeline>()
+        .init_resource::<body::HandleMap>()
+        .init_resource::<shape::HandleMap>()
+        .init_resource::<IntegrationParameters>()
+        .add_event::<CollisionEvent>()
+        .insert_resource(BroadPhase::new())
+        .insert_resource(NarrowPhase::new())
+        .insert_resource(RigidBodySet::new())
+        .insert_resource(QueryPipeline::new())
+        .insert_resource(IslandManager::new())
+        .insert_resource(ColliderSet::new())
+        .insert_resource(JointSet::new())
+        .insert_resource(CCDSolver::new())
+        .add_system_set_to_stage(self.post_physics_stage.clone(), step_systems());
```

Comment:
I don't believe this approach will work for something like `bevy_ggrs` (yet).  That library [`impl Stage for GGRSStage`](https://github.com/gschup/bevy_ggrs/blob/main/src/ggrs_stage.rs#L34), which it uses to handle updating game state with `ggrs`, which it [adds before `CoreStage::Update`](https://github.com/gschup/bevy_ggrs/blob/2348eb5f865b4cc7d565d08866dba78ad57118e3/src/lib.rs#L145). Unfortunately, [`add_system_set_to_stage` only works for SystemStages](https://github.com/bevyengine/bevy/blob/de677dbfc9dacf1724367024fc1965dae27325a4/crates/bevy_ecs/src/schedule/mod.rs#L251).  This means, specifically for `bevy_ggrs` at least, that bevy function will not find the `GGRSStage` and add the desired step systems.

That's not to say this PR doesn't work, just that it will not work with other plugins that implement custom stages (i.e., `bevy_ggrs`).  (I am guessing `bevy_backroll` also falls into this category, but I have not read much, going by this [`impl Stage`](https://github.com/HouraiTeahouse/backroll-rs/blob/main/bevy_backroll/src/lib.rs#L240))

Reply:
Interesting. Not sure if we even can do anything for it until [stageless](https://github.com/bevyengine/rfcs/pull/45).